File: backend/app/services/dashboard_service.py
from sqlalchemy import func
from sqlalchemy.orm import Session

from backend.app.database.models import PredictionAudit


def get_dashboard_stats(db: Session) -> dict:
    total_predictions = (
        db.query(func.count(PredictionAudit.id))
        .scalar()
        or 0
    )

    fraud_predictions = (
        db.query(func.count(PredictionAudit.id))
        .filter(PredictionAudit.prediction == 1)
        .scalar()
        or 0
    )

    non_fraud_predictions = (
        total_predictions - fraud_predictions
    )

    fraud_rate = (
        (fraud_predictions / total_predictions) * 100
        if total_predictions > 0
        else 0.0
    )

    average_risk_score = (
        db.query(func.avg(PredictionAudit.risk_score))
        .scalar()
        or 0.0
    )

    return {
        "total_predictions": total_predictions,
        "fraud_predictions": fraud_predictions,
        "non_fraud_predictions": non_fraud_predictions,
        "fraud_rate": round(fraud_rate, 2),
        "average_risk_score": round(
            float(average_risk_score),
            2,
        ),
    }

def get_risk_distribution(
    db: Session,
) -> dict[str, int]:

    risk_levels = (
        db.query(
            PredictionAudit.risk_level,
            func.count(PredictionAudit.id),
        )
        .group_by(PredictionAudit.risk_level)
        .all()
    )

    distribution = {
        "low": 0,
        "medium": 0,
        "high": 0,
        "critical": 0,
    }

    for risk_level, count in risk_levels:
        key = risk_level.lower()

        if key in distribution:
            distribution[key] = count

    return distribution

def get_recent_predictions(
    db: Session,
    limit: int = 10,
) -> list[PredictionAudit]:

    return list(
        db.query(PredictionAudit)
        .order_by(
            PredictionAudit.id.desc()
        )
        .limit(limit)
        .all()
    )

File: spring-backend/src/main/java/com/finguard/springbackend/service/DashboardService.java
package com.finguard.springbackend.service;

import com.finguard.springbackend.entity.PredictionAudit;
import com.finguard.springbackend.repository.PredictionAuditRepository;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class DashboardService {

    private final PredictionAuditRepository predictionAuditRepository;

    public DashboardService(PredictionAuditRepository predictionAuditRepository) {
        this.predictionAuditRepository = predictionAuditRepository;
    }

    public Map<String, Object> getDashboardStats() {

        List<PredictionAudit> audits = predictionAuditRepository.findAll();

        long totalPredictions = audits.size();

        long fraudPredictions = audits.stream()
                .filter(audit -> audit.getPrediction() == 1)
                .count();

        long nonFraudPredictions =
                totalPredictions - fraudPredictions;

        double fraudRate =
                totalPredictions > 0
                        ? ((double) fraudPredictions / totalPredictions) * 100
                        : 0.0;

        double averageRiskScore =
                audits.stream()
                        .mapToDouble(PredictionAudit::getRiskScore)
                        .average()
                        .orElse(0.0);

        Map<String, Object> result = new HashMap<>();

        result.put("total_predictions", totalPredictions);
        result.put("fraud_predictions", fraudPredictions);
        result.put("non_fraud_predictions", nonFraudPredictions);
        result.put("fraud_rate", Math.round(fraudRate * 100.0) / 100.0);
        result.put(
                "average_risk_score",
                Math.round(averageRiskScore * 100.0) / 100.0
        );

        return result;
    }

    public Map<String, Integer> getRiskDistribution() {

        List<PredictionAudit> audits =
                predictionAuditRepository.findAll();

        Map<String, Integer> distribution = new HashMap<>();

        distribution.put("low", 0);
        distribution.put("medium", 0);
        distribution.put("high", 0);
        distribution.put("critical", 0);

        for (PredictionAudit audit : audits) {

            String key =
                    audit.getRiskLevel().toLowerCase();

            if (distribution.containsKey(key)) {
                distribution.put(
                        key,
                        distribution.get(key) + 1
                );
            }
        }

        return distribution;
    }

    public List<PredictionAudit> getRecentPredictions() {

        List<PredictionAudit> audits =
                predictionAuditRepository.findAll();

        return audits.stream()
                .sorted(
                        (a, b) -> b.getCreatedAt()
                                .compareTo(a.getCreatedAt())
                )
                .limit(10)
                .toList();
    }
}